consensus/polybft/proposer_calculator.go

```diff
@@ -0,0 +1,437 @@
+	Height     uint64
+	Round      uint64
+	Proposer   *PrioritizedValidator
+	Validators []*PrioritizedValidator
+}
+
+// NewProposerSnapshotFromState create ProposerSnapshot from state if possible or from genesis block
+func NewProposerSnapshotFromState(config *runtimeConfig, logger hclog.Logger) (*ProposerSnapshot, error) {
+	snapshot, err := config.State.getProposerSnapshot()
+	if err != nil {
+		return nil, err
+	}
+
+	if snapshot == nil {
+		// pick validator set from genesis block if snapshot is not saved in db
+		genesisValidatorsSet, err := config.polybftBackend.GetValidators(0, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		snapshot = NewProposerSnapshot(1, genesisValidatorsSet)
+	}
+
+	return snapshot, nil
+}
+
+// NewProposerSnapshot creates ProposerSnapshot with height and validators with all priorities set to zero
+func NewProposerSnapshot(height uint64, validators []*ValidatorMetadata) *ProposerSnapshot {
+	validatorsSnap := make([]*PrioritizedValidator, len(validators))
+
+	for i, x := range validators {
+		validatorsSnap[i] = &PrioritizedValidator{Metadata: x, ProposerPriority: int64(0)}
+	}
+
+	return &ProposerSnapshot{
+		Round:      0,
+		Proposer:   nil,
+		Height:     height,
+		Validators: validatorsSnap,
+	}
+}
+
+// Gets total voting power from all the validators
+func (pcs ProposerSnapshot) GetTotalVotingPower() int64 {
+	totalVotingPower := int64(0)
+
+	for _, v := range pcs.Validators {
+		totalVotingPower = safeAddClip(totalVotingPower, int64(v.Metadata.VotingPower))
+	}
+
+	return totalVotingPower
+}
+
+// Returns copy of current ProposerSnapshot object
+func (pcs *ProposerSnapshot) Copy() *ProposerSnapshot {
+	var proposer *PrioritizedValidator
+
+	valCopy := make([]*PrioritizedValidator, len(pcs.Validators))
+
+	for i, val := range pcs.Validators {
+		valCopy[i] = &PrioritizedValidator{Metadata: val.Metadata.Copy(), ProposerPriority: val.ProposerPriority}
+
+		if pcs.Proposer != nil && pcs.Proposer.Metadata.Address == val.Metadata.Address {
+			proposer = valCopy[i]
+		}
+	}
+
+	return &ProposerSnapshot{
+		Validators: valCopy,
+		Height:     pcs.Height,
+		Round:      pcs.Round,
+		Proposer:   proposer,
+	}
+}
+
+// ProposerCalculator interface - proposer calculator algorithm should implement this interface
+type ProposerCalculator interface {
+	// CalcProposer calculates next proposer
+	CalcProposer(snapshot *ProposerSnapshot, round, height uint64) (types.Address, error)
+
+	// Update updates ProposerSnapshot to block with number `blockNumber`
+	Update(snapshot *ProposerSnapshot, blockNumber uint64, config *runtimeConfig) error
+
+	// GetLatestProposer returns latest calculated proposer if any
+	GetLatestProposer(snapshot *ProposerSnapshot, round, height uint64) (types.Address, bool)
+}
+
+type proposerCalculator struct {
+	// logger instance
+	logger hclog.Logger
+}
+
+// NewProposerCalculator creates a new proposer calculator object
+func NewProposerCalculator(logger hclog.Logger) *proposerCalculator {
```

Comment:
This is maybe the most important issue of the PR and it is good to discuss it.
 So this is the trouble:
1. `Syncer` and `Consensus` work in parallel. It can happen that `Syncer` finishes while `Consensus` is in the middle of a round.

2. `ConsensusRuntime` lives while blockchain lives and, therefore, we have decided to keep latest priories  as a part of `ConsensusRuntime`. The informartion is always up to date since it is modified no matter who ever first inserts a block (`Syncer` or `Consensus`)

There can be a situation when `Syncer` updates priories while `Consensus` is still running. As a consequence, `Consensus` will have wrong information.  What is more, it can happen that `Consensus` accesses priorities while `Syncer` is changing it.  To prevent that, we make a copy of data when `FSM` is created and (of course) store it in `FSM`. That copy is used only while the current instance of `FSM` exists for `Consensus`. That is why we decoupled data (priorities) from the algorithm. Everyone keeps its  own data and uses the algorithm when it is needed.